File: 3-2.py
f = open("3-input.txt", "r")
val = f.read()
f.close()

arr = val.split('\n')
print(arr)

w = len(arr[0])
h = len(arr)

def howm(ix,iy):
    x, y = 0,0
    n = 0
    while y < h-2:
        x = (x+ix)%w
        y += iy
        c = arr[y][x]
        if c == '#' : n+= 1
    return n

print(howm(1,1), howm(3,1), howm(5,1), howm(7,1), howm(1,2))

File: 3-2.hs
import System.IO

count :: Eq a => a -> [a] -> Integer
count x arr = toInteger $ length $ filter (x==) arr

tree :: [String] -> (Integer,Integer) -> Integer
tree tmap (step,down) =
    let width = toInteger $ length $ head tmap
        ns = [0..(toInteger ((length tmap)-1))]
        zipped = tail [ (mod (step*(div x down)) width, m) | (x,m) <- (zip ns tmap), (mod x down) == 0 ]
        -- zipped = tail $ zip nmod tmap
        path = [y!!(fromInteger x) | (x,y)<-zipped]
    in (count '#' path)

main :: IO()
main = do
    file <- openFile "3-input.txt" ReadMode
    contents <- hGetContents file
    let tmap = lines contents
    let ones = [(1,1), (3,1) , (5,1), (7,1), (1,2)]
    let trees = map (\y -> tree tmap y) ones
    print trees
    print $ product trees
    -- mapM_ print zipped
    -- mapM_ print path
